ui/common/src/snabbdom.ts

```diff
@@ -18,7 +18,12 @@ export function onInsert<A extends HTMLElement>(f: (element: A) => void): Hooks
   };
 }
 
-export function bind(eventName: string, f: (e: Event) => any, redraw?: Redraw, passive = true): Hooks {
+export function bind<K extends keyof GlobalEventHandlersEventMap>(
+  eventName: K,
+  f: (ev: GlobalEventHandlersEventMap[K]) => any,
```

Comment:
What is the K generic for? I would think

```
export function bind(
  eventName: keyof GlobalEventHandlersEventMap,
  f: (ev: GlobalEventHandlersEventMap[typeof eventName]) => any,
  ...
```
because if a K instance is disjoint from the keys in GlobalEventHandlersEventMap, then the function argument constraint for f will not be satisfied and it's a compile tile failure.

Reply:
Using the generic `<K>` with `extends` constrains the type to keys of GlobalEventHandlersEventMap. `K` cannot be disjoint from those keys. If you were to try `bind('schlawg', ....)` you would get
```
Argument of type '"schlawg"' is not assignable to parameter of type 'keyof GlobalEventHandlersEventMap'.
```
Using the generic, we know that e in 
```typescript
bind('submit', e => e.submitter)
```
is of type `SubmitEvent` and that it has the submitter property.
Your alternative also works, but then we lose the narrowing for the function. It is treated as the union type of all the values of GlobalEventHandlersEventMap (`UIEvent | MouseEvent | SubmitEvent ....`)

Reply:
i get it now, thanks. some wires got crossed in my thinking of extends (like it was & or something)